File: app/src/main/java/com/google/android/sambadocumentsprovider/document/DocumentMetadata.java
/*
 * Copyright 2017 Google Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.google.android.sambadocumentsprovider.document;

import android.net.Uri;
import android.provider.DocumentsContract.Document;

import androidx.annotation.Nullable;

import android.system.OsConstants;
import android.system.StructStat;
import android.text.TextUtils;
import android.util.Log;
import android.webkit.MimeTypeMap;

import com.google.android.sambadocumentsprovider.R;
import com.google.android.sambadocumentsprovider.base.DirectoryEntry;
import com.google.android.sambadocumentsprovider.nativefacade.SmbClient;
import com.google.android.sambadocumentsprovider.nativefacade.SmbDir;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

/**
 * This is a snapshot of the metadata of a seen document. It contains its SMB URI, display name,
 * access/create/modified time, size, its children etc sometime in the past. It also contains
 * the last exception thrown when querying its children.
 * <p>
 * The metadata inside this class may be fetched at different time due to Samba client API.
 */
public class DocumentMetadata {

    private static final String TAG = "DocumentMetadata";
    private static final String GENERIC_MIME_TYPE = "application/octet-stream";
    private static final Uri SMB_BASE_URI = Uri.parse("smb://");

    private final DirectoryEntry mEntry;
    private Uri mUri;

    private final AtomicReference<StructStat> mStat = new AtomicReference<>(null);
    private final AtomicReference<Map<Uri, DocumentMetadata>> mChildren = new AtomicReference<>(null);

    private final AtomicReference<Exception> mLastChildUpdateException = new AtomicReference<>(null);
    private final AtomicReference<Exception> mLastStatException = new AtomicReference<>(null);
    private long mTimeStamp;

    public DocumentMetadata(Uri uri, DirectoryEntry entry) {
        mUri = uri;
        mEntry = entry;

        mTimeStamp = System.currentTimeMillis();
    }

    public Uri getUri() {
        return mUri;
    }

    public boolean isFileShare() {
        return mEntry.getType() == DirectoryEntry.Type.FILE_SHARE;
    }

    public Integer getIconResourceId() {
        switch (mEntry.getType()) {
            case SERVER:
                return R.drawable.ic_server;
            case FILE_SHARE:
                return R.drawable.ic_folder_shared;
            default:
                // Tells SAF to use the default icon.
                return null;
        }
    }

    public Long getLastModified() {
        final StructStat stat = mStat.get();
        return (stat == null) ? null : TimeUnit.MILLISECONDS.convert(stat.st_mtime, TimeUnit.SECONDS);
    }

    public String getDisplayName() {
        return mEntry.getName();
    }

    public String getComment() {
        return mEntry.getComment();
    }

    public boolean needsStat() {
        return hasStat() && mStat.get() == null;
    }

    private boolean hasStat() {
        switch (mEntry.getType()) {
            case FILE:
                return true;
            case WORKGROUP:
            case SERVER:
            case FILE_SHARE:
            case DIR:
                // Everything is writable so no need to fetch stats for them.
                return false;
            default:
                throw new UnsupportedOperationException(
                        "Unsupported type of Samba directory entry: " + mEntry.getType());
        }
    }

    public Long getSize() {
        final StructStat stat = mStat.get();
        return (stat == null) ? null : stat.st_size;
    }

    public boolean canCreateDocument() {
        switch (mEntry.getType()) {
            case DIR:
            case FILE_SHARE:
                return true;
            case WORKGROUP:
            case SERVER:
            case FILE:
                return false;
            default:
                throw new UnsupportedOperationException(
                        "Unsupported type of Samba directory entry " + mEntry.getType());
        }
    }

    public String getMimeType() {
        switch (mEntry.getType()) {
            case FILE_SHARE:
            case WORKGROUP:
            case SERVER:
            case DIR:
                return Document.MIME_TYPE_DIR;

            case LINK:
            case COMMS_SHARE:
            case IPC_SHARE:
            case PRINTER_SHARE:
                throw new UnsupportedOperationException(
                        "Unsupported type of Samba directory entry " + mEntry.getType());

            case FILE:
                final String ext = getExtension(mEntry.getName());
                if (ext == null) {
                    return GENERIC_MIME_TYPE;
                }

                final String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
                return (mimeType == null) ? GENERIC_MIME_TYPE : mimeType;
        }

        throw new IllegalStateException("Should never reach here.");
    }

    private String getExtension(String name) {
        if (TextUtils.isEmpty(name)) {
            return null;
        }

        final int idxOfDot = name.lastIndexOf('.', name.length() - 1);
        if (idxOfDot <= 0) {
            return null;
        }

        return name.substring(idxOfDot + 1).toLowerCase();
    }

    public void reset() {
        mStat.set(null);
        mChildren.set(null);
    }

    public long getTimeStamp() {
        return mTimeStamp;
    }

    public void throwLastChildUpdateExceptionIfAny() throws Exception {
        final Exception e = mLastChildUpdateException.get();
        if (e != null) {
            mLastChildUpdateException.set(null);
            throw e;
        }
    }

    public boolean hasLoadingStatFailed() {
        final Exception e = mLastStatException.getAndSet(null);
        return e != null;
    }

    public void rename(Uri newUri) {
        mEntry.setName(newUri.getLastPathSegment());
        mUri = newUri;
    }

    /**
     * Gets children of this document.
     *
     * @return the list of children or {@code null} if it's not fetched yet.
     */
    public @Nullable
    Map<Uri, DocumentMetadata> getChildren() {
        return mChildren.get();
    }

    @Override
    public String toString() {
        return "DocumentMetadata{" +
                "mEntry=" + mEntry +
                ", mUri=" + mUri +
                ", mStat=" + mStat +
                ", mChildren=" + mChildren +
                ", mLastChildUpdateException=" + mLastChildUpdateException +
                ", mLastStatException=" + mLastStatException +
                ", mTimeStamp=" + mTimeStamp +
                '}';
    }

    public void loadChildren(SmbClient client) throws IOException {
        try (final SmbDir dir = client.openDir(mUri.toString())) {

            Map<Uri, DocumentMetadata> children = new HashMap<>();
            DirectoryEntry entry;
            while ((entry = dir.readDir()) != null) {
                Uri childUri = DocumentMetadata.buildChildUri(mUri, entry);
                if (childUri != null) {
                    children.put(childUri, new DocumentMetadata(childUri, entry));
                }
            }

            mChildren.set(children);
            mTimeStamp = System.currentTimeMillis();

        } catch (Exception e) {
            Log.e(TAG, "Failed to load children.", e);
            mLastChildUpdateException.set(e);
            throw e;
        }
    }

    public void putChild(DocumentMetadata child) {
        Map<Uri, DocumentMetadata> children = mChildren.get();
        if (children != null) {
            children.put(child.getUri(), child);
        }
    }

    void loadStat(SmbClient client) throws IOException {
        try {
            mStat.set(client.stat(mUri.toString()));

            mTimeStamp = System.currentTimeMillis();
        } catch (Exception e) {
            Log.e(TAG, "Failed to get stat.", e);
            mLastStatException.set(e);
            throw e;
        }
    }

    public static Uri buildChildUri(Uri parentUri, DirectoryEntry entry) {
        switch (entry.getType()) {
            // TODO: Support LINK type?
            case LINK:
            case COMMS_SHARE:
            case IPC_SHARE:
            case PRINTER_SHARE:
                Log.i(TAG, "Found unsupported type: " + entry.getType()
                        + " name: " + entry.getName()
                        + " comment: " + entry.getComment());
                return null;
            case WORKGROUP:
            case SERVER:
                return SMB_BASE_URI.buildUpon().authority(entry.getName()).build();
            case FILE_SHARE:
            case DIR:
            case FILE:
                return buildChildUri(parentUri, entry.getName());
        }

        Log.w(TAG, "Unknown type: " + entry.getType()
                + " name: " + entry.getName()
                + " comment: " + entry.getComment());
        return null;
    }

    public static Uri buildChildUri(Uri parentUri, String displayName) {
        if (".".equals(displayName) || "..".equals(displayName)) {
            return null;
        } else {
            return parentUri.buildUpon().appendPath(displayName).build();
        }
    }

    public static Uri buildParentUri(Uri childUri) {
        final List<String> segments = childUri.getPathSegments();
        if (segments.isEmpty()) {
            // This is possibly a server or a workgroup. We don't know its exact parent, so just return
            // "smb://".
            return SMB_BASE_URI;
        }

        Uri.Builder builder = SMB_BASE_URI.buildUpon().authority(childUri.getAuthority());
        for (int i = 0; i < segments.size() - 1; ++i) {
            builder.appendPath(segments.get(i));
        }
        return builder.build();
    }

    public static boolean isServerUri(Uri uri) {
        return uri.getPathSegments().isEmpty() && !uri.getAuthority().isEmpty();
    }

    public static boolean isShareUri(Uri uri) {
        return uri.getPathSegments().size() == 1;
    }

    public static DocumentMetadata fromUri(Uri uri, SmbClient client) throws IOException {
        final List<String> pathSegments = uri.getPathSegments();
        if (pathSegments.isEmpty()) {
            throw new UnsupportedOperationException("Can't load metadata for workgroup or server.");
        }

        final StructStat stat = client.stat(uri.toString());
        final DirectoryEntry entry = new DirectoryEntry(
                OsConstants.S_ISDIR(stat.st_mode) ? DirectoryEntry.Type.DIR : DirectoryEntry.Type.FILE,
                "",
                uri.getLastPathSegment());
        final DocumentMetadata metadata = new DocumentMetadata(uri, entry);
        metadata.mStat.set(stat);

        return metadata;
    }

    public static DocumentMetadata createShare(String host, String share) {
        final Uri uri = SMB_BASE_URI.buildUpon().authority(host).encodedPath(share).build();
        return createShare(uri);
    }

    public static DocumentMetadata createServer(Uri uri) {
        return create(uri, DirectoryEntry.Type.SERVER);
    }

    public static DocumentMetadata createShare(Uri uri) {
        return create(uri, DirectoryEntry.Type.FILE_SHARE);
    }

    private static DocumentMetadata create(Uri uri, DirectoryEntry.Type type) {
        final DirectoryEntry entry =
                new DirectoryEntry(type, "", uri.getLastPathSegment());
        return new DocumentMetadata(uri, entry);
    }
}

File: app/src/main/java/com/google/android/sambadocumentsprovider/document/DocumentMetadata.kt
/*
 * Copyright 2017 Google Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.google.android.sambadocumentsprovider.document

import android.net.Uri
import android.provider.DocumentsContract
import android.system.OsConstants
import android.system.StructStat
import android.util.Log
import android.webkit.MimeTypeMap
import com.google.android.sambadocumentsprovider.R
import com.google.android.sambadocumentsprovider.base.DirectoryEntry
import com.google.android.sambadocumentsprovider.getValue
import com.google.android.sambadocumentsprovider.nativefacade.SmbClient
import com.google.android.sambadocumentsprovider.setValue
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

/**
 * This is a snapshot of the metadata of a seen document. It contains its SMB URI, display name,
 * access/create/modified time, size, its children etc sometime in the past. It also contains
 * the last exception thrown when querying its children.
 *
 *
 * The metadata inside this class may be fetched at different time due to Samba client API.
 */
class DocumentMetadata(var uri: Uri, private val entry: DirectoryEntry) {
    private var stat by AtomicReference<StructStat?>(null)
    var children by AtomicReference<MutableMap<Uri, DocumentMetadata>?>(null)

    private val lastChildUpdateException = AtomicReference<Exception?>(null)
    private val lastStatException = AtomicReference<Exception?>(null)
    var timeStamp: Long = System.currentTimeMillis()
        private set
    val isFileShare: Boolean
        get() = entry.type == DirectoryEntry.Type.FILE_SHARE

    val iconResourceId: Int?
        get() = when (entry.type) {
            DirectoryEntry.Type.SERVER -> R.drawable.ic_server
            DirectoryEntry.Type.FILE_SHARE -> R.drawable.ic_folder_shared
            else -> null  // Tells SAF to use the default icon.
        }
    val lastModified: Long?
        get() = stat?.let { stat ->
            TimeUnit.MILLISECONDS.convert(stat.st_mtime, TimeUnit.SECONDS)
        }
    val displayName: String?
        get() = entry.name
    val comment: String
        get() = entry.comment

    fun needsStat(): Boolean {
        return hasStat() && stat == null
    }

    private fun hasStat(): Boolean {
        return when (entry.type) {
            DirectoryEntry.Type.FILE -> true
            DirectoryEntry.Type.WORKGROUP, DirectoryEntry.Type.SERVER,
            DirectoryEntry.Type.FILE_SHARE, DirectoryEntry.Type.DIR -> {
                // Everything is writable so no need to fetch stats for them.
                false
            }
            else -> throw UnsupportedOperationException(
                "Unsupported type of Samba directory entry: ${entry.type}"
            )
        }
    }

    val size: Long?
        get() = stat?.st_size

    fun canCreateDocument(): Boolean {
        return when (entry.type) {
            DirectoryEntry.Type.DIR, DirectoryEntry.Type.FILE_SHARE -> true
            DirectoryEntry.Type.WORKGROUP, DirectoryEntry.Type.SERVER,
            DirectoryEntry.Type.FILE -> false
            else -> throw UnsupportedOperationException(
                "Unsupported type of Samba directory entry ${entry.type}"
            )
        }
    }

    val mimeType: String
        get() {
            return when (entry.type) {
                DirectoryEntry.Type.FILE_SHARE, DirectoryEntry.Type.WORKGROUP,
                DirectoryEntry.Type.SERVER, DirectoryEntry.Type.DIR -> {
                    DocumentsContract.Document.MIME_TYPE_DIR
                }
                DirectoryEntry.Type.LINK, DirectoryEntry.Type.COMMS_SHARE,
                DirectoryEntry.Type.IPC_SHARE, DirectoryEntry.Type.PRINTER_SHARE -> {
                    throw UnsupportedOperationException(
                        "Unsupported type of Samba directory entry ${entry.type}"
                    )
                }
                DirectoryEntry.Type.FILE -> {
                    val ext = getExtension(entry.name) ?: return GENERIC_MIME_TYPE
                    MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext) ?: GENERIC_MIME_TYPE
                }
            }
        }

    private fun getExtension(name: String?): String? {
        if (name.isNullOrEmpty()) {
            return null
        }
        val idxOfDot = name.lastIndexOf('.', name.length - 1)
        return if (idxOfDot <= 0) null else name.substring(idxOfDot + 1).lowercase()
    }

    fun reset() {
        stat = null
        children = null
    }

    @Throws(Exception::class)
    fun throwLastChildUpdateExceptionIfAny() {
        lastChildUpdateException.getAndSet(null)?.let { throw it }
    }

    fun hasLoadingStatFailed(): Boolean {
        return lastStatException.getAndSet(null) != null
    }

    fun rename(newUri: Uri) {
        entry.name = newUri.lastPathSegment
        uri = newUri
    }

    override fun toString(): String {
        return "DocumentMetadata{entry=$entry, uri=$uri, stat=$stat, " +
                "children=${this.children}, lastChildUpdateException=$lastChildUpdateException, " +
                "lastStatException=$lastStatException, timeStamp=$timeStamp}"
    }

    @Throws(IOException::class)
    fun loadChildren(client: SmbClient) {
        try {
            client.openDir(uri.toString()).use { dir ->
                val children = ConcurrentHashMap<Uri, DocumentMetadata>()
                generateSequence { dir.readDir() }.forEach { entry ->
                    val childUri = buildChildUri(uri, entry)
                    if (childUri != null) {
                        children[childUri] = DocumentMetadata(childUri, entry)
                    }
                }
                this.children = children
                timeStamp = System.currentTimeMillis()
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load children.", e)
            lastChildUpdateException.set(e)
            throw e
        }
    }

    fun putChild(child: DocumentMetadata) {
        children?.put(child.uri, child)
    }

    @Throws(IOException::class)
    fun loadStat(client: SmbClient) {
        try {
            stat = client.stat(uri.toString())
            timeStamp = System.currentTimeMillis()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get stat.", e)
            lastStatException.set(e)
            throw e
        }
    }

    companion object {
        private const val TAG = "DocumentMetadata"
        private const val GENERIC_MIME_TYPE = "application/octet-stream"
        private val SMB_BASE_URI = Uri.parse("smb://")
        fun buildChildUri(parentUri: Uri, entry: DirectoryEntry): Uri? {
            return when (entry.type) {
                DirectoryEntry.Type.LINK, DirectoryEntry.Type.COMMS_SHARE,
                DirectoryEntry.Type.IPC_SHARE, DirectoryEntry.Type.PRINTER_SHARE -> {
                    Log.i(
                        TAG, "Found unsupported type: ${entry.type} name: ${entry.name} " +
                                "comment: ${entry.comment}"
                    )
                    null
                }
                DirectoryEntry.Type.WORKGROUP, DirectoryEntry.Type.SERVER -> {
                    SMB_BASE_URI.buildUpon().authority(entry.name).build()
                }
                DirectoryEntry.Type.FILE_SHARE, DirectoryEntry.Type.DIR, DirectoryEntry.Type.FILE -> {
                    buildChildUri(parentUri, entry.name)
                }
            }
        }

        fun buildChildUri(parentUri: Uri, displayName: String?): Uri? {
            return if ("." == displayName || ".." == displayName) {
                null
            } else {
                parentUri.buildUpon().appendPath(displayName).build()
            }
        }

        fun buildParentUri(childUri: Uri): Uri {
            val segments = childUri.pathSegments
            if (segments.isEmpty()) {
                // This is possibly a server or a workgroup. We don't know its exact parent, so just
                // return "smb://".
                return SMB_BASE_URI
            }
            val builder = SMB_BASE_URI.buildUpon().authority(childUri.authority)
            for (segment in segments) {
                builder.appendPath(segment)
            }
            return builder.build()
        }

        fun isServerUri(uri: Uri): Boolean {
            return uri.pathSegments.isEmpty() && uri.authority?.isNotEmpty() == true
        }

        fun isShareUri(uri: Uri): Boolean {
            return uri.pathSegments.size == 1
        }

        @Throws(IOException::class)
        fun fromUri(uri: Uri, client: SmbClient): DocumentMetadata {
            val pathSegments = uri.pathSegments
            if (pathSegments.isEmpty()) {
                throw UnsupportedOperationException("Can't load metadata for workgroup or server.")
            }
            val stat = client.stat(uri.toString())
            val entry = DirectoryEntry(
                if (OsConstants.S_ISDIR(stat.st_mode)) DirectoryEntry.Type.DIR else DirectoryEntry.Type.FILE,
                comment = "",
                uri.lastPathSegment
            )
            return DocumentMetadata(uri, entry).apply {
                this.stat = stat
            }
        }

        fun createShare(host: String, share: String): DocumentMetadata {
            val uri = SMB_BASE_URI.buildUpon().authority(host).encodedPath(share).build()
            return createShare(uri)
        }

        fun createServer(uri: Uri): DocumentMetadata = create(uri, DirectoryEntry.Type.SERVER)
        fun createShare(uri: Uri): DocumentMetadata = create(uri, DirectoryEntry.Type.FILE_SHARE)

        private fun create(uri: Uri, type: DirectoryEntry.Type): DocumentMetadata {
            val entry = DirectoryEntry(type, comment = "", uri.lastPathSegment)
            return DocumentMetadata(uri, entry)
        }
    }
}